src/SessionManager/ProgramSettings.cc

```diff
@@ -34,6 +37,86 @@ void applyOptionalArgument(T& val, const string& argument_name, const cxxopts::P
 }
 
 ProgramSettings::ProgramSettings(int argc, char** argv) {
+    if (argc > 1) {
+        spdlog::info("Using command-line settings");
+    }
+    ApplyCommandLineSettings(argc, argv);
+
+    const fs::path user_settings_path = fs::path(getenv("HOME")) / CARTA_USER_FOLDER_PREFIX / "backend.json";
+    const fs::path system_settings_path = "/etc/carta/backend.json";
+
+    json settings;
+    if (fs::exists(system_settings_path) && !no_system_config) {
+        settings = JSONConfigSettings(system_settings_path.string());
+        system_settings_json_exists = true;
+        spdlog::info("Reading system settings from {}", system_settings_path.string());
+    }
+
+    if (fs::exists(user_settings_path) && !no_user_config) {
+        auto user_settings = JSONConfigSettings(user_settings_path.string());
+        user_settings_json_exists = true;
+        spdlog::info("Reading user settings from {}", user_settings_path.string());
+        settings.merge_patch(user_settings); // user on top of system
+    }
+
+    if (system_settings_json_exists || user_settings_json_exists) {
+        settings.merge_patch(command_line_settings); // force command-line on top of user and sytem
+        SetSettingsFromJSON(settings);
+    }
+}
+
+json ProgramSettings::JSONConfigSettings(const std::string& json_file_path) {
+    std::ifstream ifs(json_file_path, std::ifstream::in);
+    json j;
+    try {
+        j = json::parse(ifs);
+    } catch (json::exception& err) {
+        spdlog::warn("Config file {} has problems, please check:", json_file_path);
+        spdlog::warn(err.what());
+    }
+    for (const auto& key : int_keys_map) {
+        if (j.contains(key.first) && !j[key.first].is_number_integer()) {
+            spdlog::warn("Problem in config file {}, at key {}: current value is {}, and a number is expected.", json_file_path, key.first,
+                j[key.first]);
+        }
+    }
+    for (const auto& key : bool_keys_map) {
+        if (j.contains(key.first) && !j[key.first].is_boolean()) {
+            spdlog::warn("Problem in config file {}, at key {}: current value is {}, and a boolean is expected.", json_file_path, key.first,
+                j[key.first]);
+        }
+    }
+    for (const auto& key : strings_keys_map) {
+        if (j.contains(key.first) && !j[key.first].is_string()) {
+            spdlog::warn("Problem in config file {}, at key {}: current value is {}, and a string is expected.", json_file_path, key.first,
+                j[key.first]);
+        }
+    }
+    return j;
+}
+
+void ProgramSettings::SetSettingsFromJSON(const json& j) {
+    for (const auto& key : int_keys_map) {
+        if (!j.contains(key.first) || !j[key.first].is_number_integer()) {
```

Comment:
If we're going to split this up like this, I think that we should *remove* the entries with bad types from the json object in `JSONConfigSettings`, instead of repeating the `is_type` checks here. Then we would only need to use `contains` here.